### PR TITLE
(Don't merge yet) New component for Keywhiz-client and other related changes

### PR DESCRIPTION
**Pull request by jayeshpise**
- Add a new component for Keywhiz-client which configures Keywhiz-sync on user's compute and connects to the Keywhiz server defined in the Keywhiz cloud service
- It also creates group and client on Keywhiz server
- Create new cloud service for Keywhiz server
- Also add delete recipe code on certificate cookbook

### CODE REVIEW

File: components/cookbooks/keywhiz-client/recipes/delete.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+ci_id = node.workorder.rfcCi.ciId
+
+certificate = Hash.new
+certificate["common_name"] = "keywhiz-" + ci_id.to_s
+certificate["owner_email"] = node.keywhiz_sync_cert_owner_dl
+
+node.set[:certificate] = certificate
+Chef::Log.info("common is: " + certificate["common_name"])
+include_recipe provider + "::delete_cert"
```
Comment (sureshg) [inline components/cookbooks/keywhiz-client/recipes/delete.rb]: As we discussed, the following cleanups should be done

-  Stop/Disable the keysync service
-  Delete keywhiz client from leywhiz server 
- Cleanup binary, service files and certs (`/opt/oneops/keywhiz`)

File: components/cookbooks/keywhiz-client/recipes/set_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+node.set[:keywhiz_sync_cert_domain] = meta_params["cert_domain"]
+node.set[:keywhiz_service_host] = keywhiz_service_host
+node.set[:keywhiz_service_port] = keywhiz_service_port
+
+node.set[:sync_cert_passphrase] = Array.new(4){[*("A".."Z"), *(0..9), *("a".."z")].sample}.join.to_s +
+ "-" + Array.new(4){[*("A".."Z"), *(0..9), *("a".."z")].sample}.join.to_s +
+"@" + Array.new(4){[*(0..9)].sample}.join.to_s
+#keywhiz_cloud_service[:sync_cert_passphrase]
+
+Chef::Log.info("temp log: generated paswd: " + node[:sync_cert_passphrase])
```
Comment (sureshg) [inline components/cookbooks/keywhiz-client/recipes/set_attributes.rb]: Should we need to log this password? I know this is a temp one.

File: components/cookbooks/keywhiz-cloud-service/metadata.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  :encrypted => true,
+  :required => "required",
+  :format => {
+    :category => '1.Global',
+    :order => 3,
+    :help => 'Combined client cert (in pem format including key, cert and ca cert) provided by keywhiz server for mutual TLS auth.'
+  }
+
+attribute 'keysync_download_url',
+  :description => "Keysync download url",
```
Comment (sureshg) [inline components/cookbooks/keywhiz-cloud-service/metadata.rb]: Also, mention about the file type (eg:  in our case, it should be `tar.gz`)

File: components/cookbooks/keywhiz-cloud-service/metadata.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  :required => "required",
+  :format => {
+    :category => '1.Global',
+    :order => 4,
+    :help => 'Credential app id used to access the Cert Service API'
```
Comment (sureshg) [inline components/cookbooks/keywhiz-cloud-service/metadata.rb]: The description seems wrong for download url

File: components/cookbooks/keywhiz-cloud-service/metadata.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#  :required => "required",
+  :format => {
+    :category => '1.Global',
+    :data_type => 'hash',
+#    :default => '{}',
+    :order => 5,
+    :help => 'Attributes required for configuring keysync-cert'
+  }
+
+attribute 'sync_cert_passphrase',
```
Comment (sureshg) [inline components/cookbooks/keywhiz-cloud-service/metadata.rb]: As we discussed, we should remove this attribute.

File: components/cookbooks/certificate/metadata.rb [hunk 6]
```diff
@@ -143,6 +143,17 @@
     :order => 12
   }
 
+attribute 'owner_email',
+  :description => "Application Owner Email DL (Mandatory)",
```
Comment (jayeshpise) [inline components/cookbooks/certificate/metadata.rb]: @sureshg this attribute is enabled only when "auto-generate" option is selected. That is why can not be made "required"

File: components/cookbooks/certificate/recipes/delete.rb [hunk 7]
```diff
@@ -2,3 +2,20 @@
...
 end
+
+cloud_name = node[:workorder][:cloud][:ciName]
+provider = ""
+auto_provision = node.workorder.rfcCi.ciAttributes.auto_provision
+cert_service = node[:workorder][:services][:certificate]
+
+if ! cert_service.nil? && ! cert_service[cloud_name].nil?
+        provider = node[:workorder][:services][:certificate][cloud_name][:ciClassName].gsub("cloud.service.","").downcase.split(".").last
+elsif !auto_provision.nil? && auto_provision == "true"
```
Comment (jayeshpise) [inline components/cookbooks/certificate/recipes/delete.rb]: Yeah, if not auto-generated, the certificate is just bunch of attributes/values. It is used by other cookbooks in that case - like keystore/lb

File: components/cookbooks/keywhiz-client/recipes/add.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	Chef::Log.error("Certificate cloud service not defined for this cloud")
+	exit 1
+end
+
+ci_id = node.workorder.rfcCi.ciId
+
+certificate = Hash.new
+#certificate["common_name"] = node[:oneops_server] + "-" + "keywhiz-sync-" + ci_id.to_s
+certificate["common_name"] = "keywhiz-sync-" + ci_id.to_s
+certificate["san"] = ""
```
Comment (jayeshpise) [inline components/cookbooks/keywhiz-client/recipes/add.rb]: san is not mandatory

File: components/cookbooks/keywhiz-client/recipes/set_attributes.rb (hunk 2 above)
Comment (jayeshpise) [inline components/cookbooks/keywhiz-client/recipes/set_attributes.rb]: removed. Will push the changes

File: components/cookbooks/keywhiz-cloud-service/metadata.rb (hunk 4 above)
Comment (jayeshpise) [inline components/cookbooks/keywhiz-cloud-service/metadata.rb]: corrected the description

File: components/cookbooks/keywhiz-cloud-service/metadata.rb (hunk 5 above)
Comment (jayeshpise) [inline components/cookbooks/keywhiz-cloud-service/metadata.rb]: removed

**Comment by sureshg** [comment]
Change LGTM. Merging the PR.

**Pull request merged**